### PR TITLE
add resource scheduling chapter

### PR DESCRIPTION
**Pull request by disksing**

### CODE REVIEW

**Comment by siddontang** [comment]
PTAL @tikv/pingcap-i18n

File: src/resource-scheduling/introduction.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Introduction
+
+In a distributed database environment, resource scheduling will need to meet the following requirements:
```
Comment (lilin90) [inline src/resource-scheduling/introduction.md]:
```suggestion
In a distributed database environment, resource scheduling needs to meet the following requirements:
```

File: src/resource-scheduling/introduction.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+- Keep data high available: The scheduler needs to be able to manage data redundancy too keep cluster available when some nodes fail
```
Comment (lilin90) [inline src/resource-scheduling/introduction.md]:
```suggestion
- Keeping data highly available: The scheduler needs to be able to manage data redundancy to keep the cluster available when some nodes fail.
```

File: src/resource-scheduling/introduction.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+- Balance server load: The scheduler needs to balance the load to prevent a single node from becoming a performance bottleneck for the entire system.
+- Scalability: The scheduler needs to be able to scale to thousands of nodes.
+- Fault tolerance: The scheduling process must not be stopped by the break down by single node failure.
```
Comment (lilin90) [inline src/resource-scheduling/introduction.md]:
```suggestion
- Fault tolerance: The scheduling process must not be stopped by the breaking down caused by a single node failure.
```

File: src/resource-scheduling/introduction.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+
+In TiKV cluster, resource scheduling is done by the PD. In this chapter, we will first introduce the design of two scheduling systems (Kubernetes and Mesos), followed by the design and implementation of scheduler and placement in PD.
```
Comment (lilin90) [inline src/resource-scheduling/introduction.md]:
```suggestion
In the TiKV cluster, resource scheduling is done by the PD. In this chapter, we will first introduce the design of two scheduling systems (Kubernetes and Mesos), followed by the design and implementation of scheduler and placement in PD.
```

File: src/resource-scheduling/mesos.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+![Figure 1](mesos-architecture.png)
+
+## Components
+
+Mesos consists of a master process that manages slave daemons running on each cluster node, and frameworks that run tasks on these slaves.
+
+- Mesos master
+
+The _master_ sees the global information, and is responsible for resource scheduling and logical control between different _frameworks_. The _frameworks_ need to be registered to _master_ in order to be used. It uses Zookeeper to achieve HA.
```
Comment (lilin90) [inline src/resource-scheduling/mesos.md]:
```suggestion
    The _master_ sees the global information, and is responsible for resource scheduling and logical control between different _frameworks_. The _frameworks_ need to be registered to _master_ in order to be used. It uses Zookeeper to achieve HA.
```

File: src/resource-scheduling/mesos.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+- Mesos salve
+
+The _slave_ is responsible for reporting the resource status (idle resources, running status, etc.) on the slave node to _master_, and is responsible for isolating the local resources to perform the specific tasks assigned by master.
```
Comment (lilin90) [inline src/resource-scheduling/mesos.md]:
```suggestion
    The _slave_ is responsible for reporting the resource status (idle resources, running status, etc.) on the slave node to _master_, and is responsible for isolating the local resources to perform the specific tasks assigned by master.
```

File: src/resource-scheduling/mesos.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+- Frameworks
+
+Each _framework_ consists of two components: a _scheduler_ that registers with the _master_ to be offered resources, and an _executor_ process that is launched on _slave_ nodes to run the _framework_’s tasks.
```
Comment (lilin90) [inline src/resource-scheduling/mesos.md]:
```suggestion
    Each _framework_ consists of two components: a _scheduler_ that registers with the _master_ to be offered resources, and an _executor_ process that is launched on _slave_ nodes to run the _framework_’s tasks.
```

File: src/resource-scheduling/mesos.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+## Resource scheduling
+To support the sophisticated schedulers of today's frameworks, Mesos introduces a distributed two-level scheduling mechanism called _resource offers_.
+
+Each resource offer is a list of free resources (for example, <1Core CPU, 2GB RAM>) on multiple slaves. While the _master_ decides how many resources to offer to each framework according to an organizational policy, the frameworks’ schedulers select which of the offered resources to use. When a framework accepts offered resources, it passes Mesos a description of the tasks it wants to launch on them.
+
+![Figure 2](mesos-scheduling.png)
+
+The figure shows an example of how resource scheduling works:
+
+- Slave 1 reports to the master that it has 4 CPUs and 4 GB of memory free. The master then invokes the allocation policy module, which tells it that framework 1 should be offered all available resources.
```
Comment (lilin90) [inline src/resource-scheduling/mesos.md]: I think an ordered list might be more appropriate for this part.

File: src/resource-scheduling/scheduler-of-kubernetes.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
+# Scheduler of Kubernetes
+
+## Ovewview
```
Comment (lilin90) [inline src/resource-scheduling/scheduler-of-kubernetes.md]:
```suggestion
## Overview
```

File: src/resource-scheduling/scheduler-of-kubernetes.md [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+Kubernetes built on top of Docker can build a container scheduling service, which is designed to allow users to manage cloud container clusters through Kubernetes clusters without the need for complex setup tasks. The system will automatically select the appropriate working node to perform specific container cluster scheduling processing.
+
+The scheduler needs to take into account individual and collective resource requirements, quality of service requirements, hardware/software/policy constraints, affinity and anti-affinity specifications, data locality, inter-workload interference, deadlines, and so on.
+
+## Scheduling process
+
+The scheduling process is mainly divided into 2 steps. In the _predicate_ step, the scheduler filters out nodes that do not satisfy required conditions. And in the _priority_ step, the scheduler sorts the nodes that meet all of the fit predicates, and then chooses the best one.
+
+### predicate stage
```
Comment (lilin90) [inline src/resource-scheduling/scheduler-of-kubernetes.md]:
```suggestion
### Predicate stage
```

File: src/resource-scheduling/scheduler-of-kubernetes.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+Kubernetes scheduler provides some predicates algorithms by default. For instance, the `HostNamePred` pridicate checks if the host name matches requested hostname; the `PodsFitsResourcePred` checks if a node has sufficient resources, such as cpu, memory, gpu, opaque int resources etc to run a pod. Relevant code can be found in [kubernetes/pkg/scheduler/algorithm/predicates/](https://github.com/kubernetes/kubernetes/tree/master/pkg/scheduler/algorithm/predicates).
```
Comment (lilin90) [inline src/resource-scheduling/scheduler-of-kubernetes.md]:
```suggestion
Kubernetes scheduler provides some predicates algorithms by default. For instance, the `HostNamePred` predicate checks if the hostname matches the requested hostname; the `PodsFitsResourcePred` checks if a node has sufficient resources, such as CPU, memory, GPU, opaque int resources and so on, to run a pod. Relevant code can be found in [kubernetes/pkg/scheduler/algorithm/predicates/](https://github.com/kubernetes/kubernetes/tree/master/pkg/scheduler/algorithm/predicates).
```

File: src/resource-scheduling/scheduler-of-kubernetes.md [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+### priority stage
```
Comment (lilin90) [inline src/resource-scheduling/scheduler-of-kubernetes.md]:
```suggestion
### Priority stage
```

File: src/resource-scheduling/scheduler-of-kubernetes.md [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+In the priority step, the scheduler uses `PrioritizeNodes` function to rank all nodes by calling each priority funcitons sequentially:
```
Comment (lilin90) [inline src/resource-scheduling/scheduler-of-kubernetes.md]:
```suggestion
In the priority step, the scheduler uses the `PrioritizeNodes` function to rank all nodes by calling each priority functions sequentially:
```

File: src/resource-scheduling/scheduler-of-kubernetes.md [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+- Each priority function is expected to set a score of 0-10 where 0 is the lowest priority score (least preferred node) and 10 is the highest.
+- Add all (weighted) scores for each node to get a total score.
+- Select the node with highest score.
```
Comment (lilin90) [inline src/resource-scheduling/scheduler-of-kubernetes.md]:
```suggestion
- Select the node with the highest score.
```

**Comment by disksing** [comment]
All comments addressed /cc @lilin90

**Review by lilin90 — COMMENTED** [review]
LGTM

File: src/resource-scheduling/introduction.md [hunk 15]
```diff
@@ -0,0 +1,9 @@
+# Introduction
+
+In a distributed database environment, resource scheduling needs to meet the following requirements:
+- Keeping data highly available: The scheduler needs to be able to manage data redundancy to keep the cluster available when some nodes fail.
+- Balance server load: The scheduler needs to balance the load to prevent a single node from becoming a performance bottleneck for the entire system.
+- Scalability: The scheduler needs to be able to scale to thousands of nodes.
+- Fault tolerance: The scheduling process must not be stopped by the breaking down caused by a single node failure.
+
+In the TiKV cluster, resource scheduling is done by the PD. In this chapter, we will first introduce the design of two scheduling systems (Kubernetes and Mesos), followed by the design and implementation of scheduler and placement in PD.
```
Comment (Hoverbear) [inline src/resource-scheduling/introduction.md]:
```suggestion
In the TiKV cluster, resource scheduling is done by the Placement Driver (PD). In this chapter, we will first introduce the design of two scheduling systems (Kubernetes and Mesos), followed by the design and implementation of scheduler and placement in PD.
```

File: src/resource-scheduling/mesos.md [hunk 16]
```diff
@@ -0,0 +1,58 @@
+# Mesos
+
+## Overview
+
+Mesos was originally launched by UC Berkeley's AMPLab in 2009. It is lisenced under Apache and now operated by Mesosphere, Inc.
+
+Mesos can abstract and schedule the resources of the entire data center (including CPU, memory, storage, network, etc.), so that multiple applications can run in the cluster at the same time, and do not need to care about the physical distribution of resources.
```
Comment (Hoverbear) [inline src/resource-scheduling/mesos.md]:
```suggestion
Mesos can abstract and schedule the resources of the entire data center (including CPU, memory, storage, network, etc.). This allows multiple applications to run in a cluster at the same time without needing to care about the physical distribution of resources.
```

File: src/resource-scheduling/mesos.md [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+
+Mesos has many compelling features, including:
+
+- Supports large scale scenarios with tens of thousands of nodes (adopted by Apple, Twitter, eBay, etc.)
```
Comment (Hoverbear) [inline src/resource-scheduling/mesos.md]:
```suggestion
- Support for large scale scenarios with tens of thousands of nodes (adopted by Apple, Twitter, eBay, etc.)
```

File: src/resource-scheduling/mesos.md [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+- Supports multiple application frameworks, including Marathon, Singularity, Aurora, etc.
```
Comment (Hoverbear) [inline src/resource-scheduling/mesos.md]:
```suggestion
- Support for multiple application frameworks, including Marathon, Singularity, Aurora, etc.
```

File: src/resource-scheduling/mesos.md [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+- High Available (rely on ZooKeeper)
+- Supports Docker, LXC and other container techniques
```
Comment (Hoverbear) [inline src/resource-scheduling/mesos.md]:
```suggestion
- Support for Docker, LXC and other container techniques
```

File: src/resource-scheduling/mesos.md [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+
+Mesos was originally launched by UC Berkeley's AMPLab in 2009. It is lisenced under Apache and now operated by Mesosphere, Inc.
+
+Mesos can abstract and schedule the resources of the entire data center (including CPU, memory, storage, network, etc.), so that multiple applications can run in the cluster at the same time, and do not need to care about the physical distribution of resources.
+
+Mesos has many compelling features, including:
+
+- Supports large scale scenarios with tens of thousands of nodes (adopted by Apple, Twitter, eBay, etc.)
+- Supports multiple application frameworks, including Marathon, Singularity, Aurora, etc.
+- High Available (rely on ZooKeeper)
```
Comment (Hoverbear) [inline src/resource-scheduling/mesos.md]:
```suggestion
- High Availability (relies on ZooKeeper)
```

File: src/resource-scheduling/mesos.md [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+- Supports Docker, LXC and other container techniques
+- Provides API for several popular languages, including Python, Java, C++, etc.
+- Comes with a simple and easy-to-use WebUI
```
Comment (Hoverbear) [inline src/resource-scheduling/mesos.md]:
```suggestion
- A simple and easy-to-use WebUI
```

File: src/resource-scheduling/mesos.md [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+
+Mesos can abstract and schedule the resources of the entire data center (including CPU, memory, storage, network, etc.), so that multiple applications can run in the cluster at the same time, and do not need to care about the physical distribution of resources.
+
+Mesos has many compelling features, including:
+
+- Supports large scale scenarios with tens of thousands of nodes (adopted by Apple, Twitter, eBay, etc.)
+- Supports multiple application frameworks, including Marathon, Singularity, Aurora, etc.
+- High Available (rely on ZooKeeper)
+- Supports Docker, LXC and other container techniques
+- Provides API for several popular languages, including Python, Java, C++, etc.
```
Comment (Hoverbear) [inline src/resource-scheduling/mesos.md]:
```suggestion
- Providing APIs for several popular languages, including Python, Java, C++, etc.
```

File: src/resource-scheduling/mesos.md [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+- Comes with a simple and easy-to-use WebUI
+
+## Architecture
+
+It is important to notice that Mesos itself is a resource scheduling framework. It is not a complete application management platform, so Mesos can't work only on its own. However, based on Mesos, it is relatively easy to provide distributed operation capabilities for various application management frameworks or middleware platforms. Multiple frameworks can also run in a single Mesos cluster at the same time, improving overall resource utilization efficiency.
```
Comment (Hoverbear) [inline src/resource-scheduling/mesos.md]:
```suggestion
It is important to notice that Mesos itself is only a resource scheduling framework. It is not a complete application management platform, so Mesos can't work only on its own. However, based on Mesos, it is relatively easy to provide distributed operation capabilities for various application management frameworks or middleware platforms. Multiple frameworks can also run in a single Mesos cluster at the same time, improving overall resource utilization efficiency.
```

File: src/resource-scheduling/scheduler-of-kubernetes.md [hunk 24]
```diff
@@ -0,0 +1,30 @@
+# Scheduler of Kubernetes
+
+## Overview
+
+Kubernetes is a Docker-based open source container cluster management system initiated and maintained by the Google team. It supports not only common cloud platforms but also internal data centers.
+
+Kubernetes built on top of Docker can build a container scheduling service, which is designed to allow users to manage cloud container clusters through Kubernetes clusters without the need for complex setup tasks. The system will automatically select the appropriate working node to perform specific container cluster scheduling processing.
```
Comment (Hoverbear) [inline src/resource-scheduling/scheduler-of-kubernetes.md]:
```suggestion
Kubernetes built a container scheduling service which is designed to allow users to manage cloud container clusters through Kubernetes clusters without the need for complex setup tasks. The system will automatically select the appropriate working node to perform specific container cluster scheduling processing.
```

File: src/resource-scheduling/scheduler-of-kubernetes.md [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+
+The scheduler needs to take into account individual and collective resource requirements, quality of service requirements, hardware/software/policy constraints, affinity and anti-affinity specifications, data locality, inter-workload interference, deadlines, and so on.
+
+## Scheduling process
+
+The scheduling process is mainly divided into 2 steps. In the _predicate_ step, the scheduler filters out nodes that do not satisfy required conditions. And in the _priority_ step, the scheduler sorts the nodes that meet all of the fit predicates, and then chooses the best one.
+
+### Predicate stage
+
+Kubernetes scheduler provides some predicates algorithms by default. For instance, the `HostNamePred` predicate checks if the hostname matches the requested hostname; the `PodsFitsResourcePred` checks if a node has sufficient resources, such as CPU, memory, GPU, opaque int resources and so on, to run a pod. Relevant code can be found in [kubernetes/pkg/scheduler/algorithm/predicates/](https://github.com/kubernetes/kubernetes/tree/master/pkg/scheduler/algorithm/predicates).
```
Comment (Hoverbear) [inline src/resource-scheduling/scheduler-of-kubernetes.md]:
```suggestion
The scheduler provides some predicates algorithms by default. For instance, the `HostNamePred` predicate checks if the hostname matches the requested hostname; the `PodsFitsResourcePred` checks if a node has sufficient resources, such as CPU, memory, GPU, opaque int resources and so on, to run a pod. Relevant code can be found in [kubernetes/pkg/scheduler/algorithm/predicates/](https://github.com/kubernetes/kubernetes/tree/master/pkg/scheduler/algorithm/predicates).
```

File: src/resource-scheduling/scheduler-of-kubernetes.md [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+
+### Priority stage
+
+In the priority step, the scheduler uses the `PrioritizeNodes` function to rank all nodes by calling each priority functions sequentially:
+- Each priority function is expected to set a score of 0-10 where 0 is the lowest priority score (least preferred node) and 10 is the highest.
+- Add all (weighted) scores for each node to get a total score.
+- Select the node with the highest score.
+
+## References
+1. [Kubernetes document](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/)
```
Comment (Hoverbear) [inline src/resource-scheduling/scheduler-of-kubernetes.md]:
```suggestion
1. [`kube-scheduler` documentation](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/)
```

**Comment by Hoverbear** [comment]
Nice job. :)

Can you please confirm that we have the right to use these images? Did you make them? Are they CC licensed so we can just link to the source?

**Comment by disksing** [comment]
@Hoverbear Thanks for reminding. Actually, I'm not sure how these images are licensed. I took them from the mesos paper.

**Comment by Hoverbear** [comment]
@disksing Ok! We need to either make new versions, find new versions, ask Mesos to let us use theirs, or remove them. What do you prefer? I'm happy to remake them if you don't have time.

**Comment by disksing** [comment]
@Hoverbear I redrew the images, please take another look.

**Review by Hoverbear — APPROVED** [review]

**Pull request merged**